### PR TITLE
[workflow] add openstack_workflow_workflow_v2 datasource

### PR DESCRIPTION
**Pull request by anthonyroussel**
See issue #1818

Depends on https://github.com/gophercloud/utils/pull/234 to add the Mistral Client

I propose to start adding Mistral support to the Terraform OpenStack provider, starting with the workflow datasource. The workflow resource will be a little harder to implement first, since the POST /v2/workflows API doesn't return the ID of the newly created workflow (neither in the headers, nor in the response body).

### CODE REVIEW

**Comment by anthonyroussel** [comment]
Bobcat Mistral workflow is failing and I don't understand why yet:

```console
 Nov 04 19:53:52 fv-az1490-394 mistral-server[71449]: 2024-11-04 19:53:52.089 71449 ERROR oslo_service.service     raise exceptions.Etcd3Exception(resp.text, resp.reason)
Nov 04 19:53:52 fv-az1490-394 mistral-server[71449]: 2024-11-04 19:53:52.089 71449 ERROR oslo_service.service etcd3gw.exceptions.Etcd3Exception: Not Found
Nov 04 19:53:52 fv-az1490-394 mistral-server[71449]: 2024-11-04 19:53:52.089 71449 ERROR oslo_service.service 
Nov 04 19:53:52 fv-az1490-394 mistral-server[71449]: 2024-11-04 19:53:52.090 71449 INFO mistral.engine.engine_server [-] Stopping an engine server [graceful=True, timeout=60]
```

See https://github.com/anthonyroussel/terraform-provider-openstack/actions/runs/11671350585/job/32498046218

**Comment by anthonyroussel** [comment]
## Failing 2023.2 Mistral pipeline
Fixed the Mistral pipeline by integrating these 2 patches for the 2023.2 (bobcat) workflow:
* https://github.com/openstack/mistral/commit/e343ccb078d8ba261ac70afca93f4358589730d3
* https://github.com/openstack/mistral/commit/ecdeadeb7a1aa87cba2cdb0c1a2bb1ffc4aabf25

Mistral Devstack's plugin version has been set to commit `ecdeadeb7a1aa87cba2cdb0c1a2bb1ffc4aabf25` (stable/2023.2 + release commit + these 2 patches) for the Mistral workflow.

## Gophercloud PR
The gophercloud/utils PR to add the WorkflowV2Client to utils has been merged, see https://github.com/gophercloud/utils/pull/234

File: .github/workflows/functional-workflow.yml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            ubuntu_version: "22.04"
+            mistral_plugin_version: "stable/2024.1"
+          - name: "bobcat"
+            openstack_version: "stable/2023.2"
+            ubuntu_version: "22.04"
+            # Devstack support is broken with Mistral v2023.2, and requires 2 patches:
+            # * https://github.com/openstack/mistral/commit/e343ccb078d8ba261ac70afca93f4358589730d3
+            # * https://github.com/openstack/mistral/commit/ecdeadeb7a1aa87cba2cdb0c1a2bb1ffc4aabf25
+            mistral_plugin_version: "ecdeadeb7a1aa87cba2cdb0c1a2bb1ffc4aabf25"
+          - name: "antelope"
```
Comment (nikParasyr) [inline .github/workflows/functional-workflow.yml:40]: Can you remove `antelope`? In general we support 3 latest + master, do with dalmatian in, antelope should be out.

File: docs/data-sources/workflow_v2.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+## Argument Reference
+
+* `region` - (Optional) The region in which to obtain the V2 Workflow client.
+
+* `name` - (Optional) The name of the workflow.
+
+* `namespace` - (Optional) The namespace of the workflow.
+
+* `project_id` - (Optional) The id of the project to retrieve the workflow.
```
Comment (nikParasyr) [inline docs/data-sources/workflow_v2.md:30]: Doesn't defining this require admin credentials? if so, please document it.

File: openstack/data_source_openstack_workflow_v2.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+			"namespace": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+
+			"input": {
+				Type:     schema.TypeString,
+				Optional: true,
```
Comment (nikParasyr) [inline openstack/data_source_openstack_workflow_v2.go:39]: From my understanding, this cannot be set by the user, thus `Optional` should be removed and only computed left.

File: openstack/data_source_openstack_workflow_v2.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+				Computed: true,
+			},
+
+			"definition": {
+				Type:     schema.TypeString,
+				Optional: true,
```
Comment (nikParasyr) [inline openstack/data_source_openstack_workflow_v2.go:45]: From my understanding, this cannot be set by the user, thus `Optional` should be removed and only computed left.

File: openstack/data_source_openstack_workflow_v2.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+				Computed: true,
+			},
+
+			"tags": {
+				Type:     schema.TypeSet,
+				Optional: true,
```
Comment (nikParasyr) [inline openstack/data_source_openstack_workflow_v2.go:51]: From my understanding, this cannot be set by the user, thus `Optional` should be removed and only computed left.

File: openstack/data_source_openstack_workflow_v2.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Set:      schema.HashString,
+			},
+
+			"scope": {
+				Type:     schema.TypeString,
+				Optional: true,
```
Comment (nikParasyr) [inline openstack/data_source_openstack_workflow_v2.go:58]: From my understanding, this cannot be set by the user, thus `Optional` should be removed and only computed left.

File: openstack/data_source_openstack_workflow_v2.go [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+			"project_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+
+			"created_at": {
+				Type:     schema.TypeString,
+				Optional: true,
```
Comment (nikParasyr) [inline openstack/data_source_openstack_workflow_v2.go:70]: From my understanding, this cannot be set by the user, thus `Optional` should be removed and only computed left.

File: openstack/provider.go [hunk 8]
```diff
@@ -312,6 +312,7 @@ func Provider() *schema.Provider {
 			"openstack_keymanager_secret_v1":                     dataSourceKeyManagerSecretV1(),
 			"openstack_keymanager_container_v1":                  dataSourceKeyManagerContainerV1(),
 			"openstack_loadbalancer_flavor_v2":                   dataSourceLBFlavorV2(),
+			"openstack_workflow_v2":                              dataSourceWorkflowV2(),
```
Comment (nikParasyr) [inline openstack/provider.go:315]: This should be `openstack_workflow_workflow_v2`

**Review by nikParasyr — CHANGES_REQUESTED** [review]
Hello. Thanks for working on this.
First review, ill need to go through some stuff again as it is quite a big PR.
Some changes requested already.

Some notes/questions:
- data source should be named `openstack_workflow_workflow_v2`. While this might feel duplication the convention is `cloud/provider_serviceName_resourceName, and we shouldnt break it
- would like input from @kayrus as well in general
- @anthonyroussel are you intending to work on more data-sources/resources of mistral? Would be good if we get an issue with all the data-sources/resources to keep track of progress.

**Comment by anthonyroussel** [comment]
Hello @nikParasyr, and thanks for the review!

> data source should be named openstack_workflow_workflow_v2. While this might feel duplication the convention is `cloud/provider_serviceName_resourceName, and we shouldnt break it

Sure! I've modified the datasource name to openstack_workflow_workflow_v2.

> are you intending to work on more data-sources/resources of mistral? Would be good if we get an issue with all the data-sources/resources to keep track of progress.

Yes, I plan to work on additional Mistral resources and datasources. I created issue #1818 to track the progress on the Mistral implementation in the provider.

File: .github/workflows/functional-workflow.yml (hunk 1 above)
Comment (anthonyroussel) [reply .github/workflows/functional-workflow.yml]: I removed antelope from build matrix.

File: openstack/provider.go (hunk 8 above)
Comment (anthonyroussel) [reply openstack/provider.go]: Sure, I renamed the datasource to openstack_workflow_workflow_v2.

File: openstack/data_source_openstack_workflow_v2.go (hunk 3 above)
Comment (anthonyroussel) [reply openstack/data_source_openstack_workflow_v2.go]: Yes you're right, I removed the `Optional` and keep the `Computed` only for these fields.

**Pull request merged**